routes/queryargs.go

```diff
@@ -0,0 +1,189 @@
+//   Copyright 2017 MSolution.IO
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package routes
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+)
+
+type (
+	// QueryArgInt is a type allowing to know the expected int
+	// type for the argument. QueryArgInt can be interfaced to
+	// QueryArgType.
+	QueryArgInt struct{}
+
+	// QueryArgUint is a type allowing to know the expected uint
+	// type for the argument. QueryArgUint can be interfaced to
+	// QueryArgType.
+	QueryArgUint struct{}
+
+	// QueryArgString is a type allowing to know the expected string
+	// type for the argument. QueryArgString can be interfaced to
+	// QueryArgType.
+	QueryArgString struct{}
+
+	// QueryArgIntSlice is a type allowing to know the expected []int
+	// type for the argument. QueryArgIntSlice can be interfaced to
+	// QueryArgType.
+	QueryArgIntSlice struct{}
+
+	// QueryArgUintSlice is a type allowing to know the expected []uint
+	// type for the argument. QueryArgUintSlice can be interfaced to
+	// QueryArgType.
+	QueryArgUintSlice struct{}
+
+	// QueryArgStringSlice is a type allowing to know the expected []string
+	// type for the argument. QueryArgStringSlice can be interfaced to
+	// QueryArgType.
+	QueryArgStringSlice struct{}
+
+	// QueryArgType is an interface used by all the type above. The Parse
+	// method takes the raw value of the argument and returns its typed value.
+	// An error can be returned if the value could not be parse. The error's
+	// message contains %s which has to be replaced by the argument's name
+	// before being displayed.
+	QueryArgType interface {
+		Parse(string) (interface{}, error)
+	}
+
+	// QueryArg is a structure defining an argument by its name and its type.
+	QueryArg struct {
+		Name string
+		Type QueryArgType
+	}
+
+	// WithQueryArg is a slice of QueryArg. That is the type that contains all
+	// the arguments to parse in the URL. WithQueryArg has a method Decorate
+	// called to apply the decorators on an endpoint.
+	WithQueryArg []QueryArg
+)
+
+// Parse is the method of QueryArgInt allowing to type the row value to
+// an int value. It can return an error, the error's message contains %s
+// which has to be replaced by the argument's name before being displayed.
+func (d QueryArgInt) Parse(val string) (interface{}, error) {
+	if i, err := strconv.ParseInt(val, 10, 64); err == nil {
```

Comment:
`strconv.ParseInt` returns an `int64`, yet this is the implementation for `QueryArgInt`. This is misleading. Please either convert the result to an `int`, or rename the type to `QueryArgInt64`.
Same with `uint`/`uint64`, `[]int`/`[]int64` and `[]uint`/`[]uint64`.

If you choose to do the conversion, consider an overflow as an error.